### PR TITLE
Cow-box WebSocketFrame.

### PR DESCRIPTION
**Pull request by marlimox**
### Motivation:

WebSocketFrameDecoder makes an allocation due to WebSocketFrame not fitting in an
existential. We should reduce this overhead.

### Modifications:

Moved data and extensionData into a cow-box and reorder fields to more efficiently
pack them.

### Result:

WebSocketFrame is now 14 bytes, down from 55.

### CODE REVIEW

File: Sources/NIOWebSocket/WebSocketFrame.swift [hunk 1]
```diff
@@ -170,30 +183,39 @@ public struct WebSocketFrame {
...
+        }
+    }
 
     /// The extension data, if any.
     ///
     /// On frames received from the network, this data is not necessarily unmasked. This is to provide as much
     /// information as possible. If unmasked data is desired, either use the computed `unmaskedExtensionData` property to
     /// obtain it, or transform this data directly by calling `extensionData.unmask(maskKey)`.
-    public var extensionData: ByteBuffer? = nil
+    public var extensionData: ByteBuffer? = nil {
```
Comment (Lukasa) [inline Sources/NIOWebSocket/WebSocketFrame.swift]:
```suggestion
    public var extensionData: ByteBuffer? {
```

File: Sources/NIOWebSocket/WebSocketFrame.swift [hunk 2]
```diff
@@ -270,3 +292,22 @@ public struct WebSocketFrame {
...
+    fileprivate class _Storage {
+        var data: ByteBuffer
+        var extensionData: ByteBuffer?
+
+        fileprivate init(data: ByteBuffer, extensionData: ByteBuffer?) {
+            self.data = data
+            self.extensionData = extensionData
+        }
+
+        fileprivate init(copying original: Self) {
```
Comment (Lukasa) [inline Sources/NIOWebSocket/WebSocketFrame.swift]:
```suggestion
        fileprivate init(copying original: WebSocketFrame._Storage) {
```

**Comment by Lukasa** [comment]
Hey @marlimox, great work! We need to drop the threshold on the allocation counter test now, as it wants us to lower the watermark for allocations so we don’t regress it.

**Comment by marlimox** [comment]
Ah yep.

**Comment by marlimox** [comment]
Thank you for the help! 😄

**Pull request merged**